hyperspy/learn/rpca.py

```diff
@@ -0,0 +1,389 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 The HyperSpy developers
+#
+# This file is part of  HyperSpy.
+#
+#  HyperSpy is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+#  HyperSpy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with  HyperSpy.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+
+import numpy as np
+import scipy.linalg
+
+from hyperspy.misc.machine_learning.import_sklearn import (
+    fast_svd, sklearn_installed)
+
+_logger = logging.getLogger(__name__)
+
+def _thresh(X, lambda1):
+    res = np.abs(X) - lambda1
+    return np.sign(X) * ((res > 0) * res)
+
+def rpca(X, rank, fast=False, lambda1=None,
+         power=None, tol=None, maxiter=None):
+    """
+    This function performs Robust PCA with missing or corrupted data,
+    using the GoDec algorithm
+
+    Parameters
+    ----------
+    X : numpy array
+        is the [nfeatures x nsamples] matrix of observations.
+    rank : int
+        The model dimensionality.
+    lambda1 : None | float
+        Regularization parameter.
+        If None, set to 1 / sqrt(nsamples)
+    power : None | integer
+        The number of power iterations used in the initialization
+        If None, set to 0 for speed
+    tol : None | float
+        Convergence tolerance
+        If None, set to 1e-3
+    maxiter : None | integer
+        Maximum number of iterations
+        If None, set to 1e3
+
+    Returns
+    -------
+    Xhat : numpy array
+        is the [nfeatures x nsamples] low-rank matrix
+    Ehat : numpy array
+        is the [nfeatures x nsamples] sparse error matrix
+    Ghat : numpy array
+        is the [nfeatures x nsamples] Gaussian noise matrix
+    U, S, V : numpy arrays
+        are the results of an SVD on Xhat
+
+    Notes
+    -----
+    Algorithm based on the following research paper:
+       Tianyi Zhou and Dacheng Tao, "GoDec: Randomized Low-rank & Sparse Matrix
+       Decomposition in Noisy Case", ICML-11, (2011), pp. 33-40.
+
```

Comment:
If this is a translation from MATLAB code, could you provide a link to the original code?